### ISSUE TITLE
custom_exit and custom_stoploss

### ISSUE THREAD

**Issue by zhihuba** [freqtrade/freqtrade#9302]
# environment info
- Python 3.11.4
- freqtrade 2023.9-dev-c894843d0
- os mac m1
# log
```bash
/Users/root/miniconda3/envs/freqtrade/bin/python /Users/root/Desktop/source_code/freqtrade/freqtrade/main.py plot-dataframe -p BTC/USDT:USDT --strategy DivergenceV5 -i 5m --datadir user_data/data/binance/ --timerange=20230902- 
2023-10-14 23:32:00,658 - freqtrade - INFO - freqtrade 2023.9-dev-c894843d0
2023-10-14 23:32:00,666 - freqtrade.configuration.load_config - INFO - Using config: user_data/config.json ...
2023-10-14 23:32:00,667 - freqtrade.loggers - INFO - Verbosity set to 0
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Parameter -i/--timeframe detected ... Using timeframe: 5m ...
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Parameter --timerange detected: 20230902- ...
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Using user-data directory: /Users/root/Desktop/source_code/freqtrade/user_data ...
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Using data directory: user_data/data/binance ...
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Overriding timeframe with Command line argument
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Using pairs ['BTC/USDT:USDT']
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Limiting plot to: 750
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Using trades from: file
2023-10-14 23:32:00,667 - freqtrade.configuration.configuration - INFO - Filter trades by timerange: 20230902-
2023-10-14 23:32:00,668 - freqtrade.exchange.check_exchange - INFO - Checking exchange...
2023-10-14 23:32:00,671 - freqtrade.exchange.check_exchange - INFO - Exchange "binance" is officially supported by the Freqtrade development team.
2023-10-14 23:32:00,671 - freqtrade.configuration.config_validation - INFO - Validating configuration ...
2023-10-14 23:32:00,769 - freqtrade.resolvers.iresolver - INFO - Using resolved strategy DivergenceV5 from '/Users/root/Desktop/source_code/freqtrade/user_data/strategies/DivergenceV5.py'...
2023-10-14 23:32:00,769 - freqtrade.strategy.hyper - INFO - Loading parameters from file /Users/root/Desktop/source_code/freqtrade/user_data/strategies/DivergenceV5.json
2023-10-14 23:32:00,769 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'timeframe' with value in config file: 5m.
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_currency' with value in config file: USDT.
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_amount' with value in config file: unlimited.
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'unfilledtimeout' with value in config file: {'entry': 10, 'exit': 10, 'exit_timeout_count': 0, 'unit': 'minutes'}.
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'max_open_trades' with value in config file: 1.
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using minimal_roi: {'0': 0.099, '40': 0.029, '76': 0.011, '165': 0}
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using timeframe: 5m
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stoploss: -0.333
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop: True
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive: 0.13
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive_offset: 0.181
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_only_offset_is_reached: True
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_custom_stoploss: True
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using process_only_new_candles: True
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_types: {'entry': 'limit', 'exit': 'limit', 'stoploss': 'market', 'stoploss_on_exchange': False}
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_time_in_force: {'entry': 'GTC', 'exit': 'GTC'}
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_currency: USDT
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_amount: unlimited
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using protections: []
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using startup_candle_count: 50
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using unfilledtimeout: {'entry': 10, 'exit': 10, 'exit_timeout_count': 0, 'unit': 'minutes'}
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_exit_signal: True
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using exit_profit_only: False
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_roi_if_entry_signal: False
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using exit_profit_offset: 0.0
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using disable_dataframe_checks: False
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_buying_expired_candle_after: 0
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using position_adjustment_enable: False
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using max_entry_position_adjustment: -1
2023-10-14 23:32:00,770 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using max_open_trades: 1
2023-10-14 23:32:00,770 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2023-10-14 23:32:00,770 - freqtrade.exchange.exchange - INFO - Using CCXT 4.0.76
2023-10-14 23:32:00,770 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'options': {'defaultType': 'swap'}}
2023-10-14 23:32:00,773 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'options': {'defaultType': 'swap'}}
2023-10-14 23:32:00,776 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2023-10-14 23:32:04,539 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange 'Binance'...
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - Strategy Parameter: timeperiod = 30
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - Strategy Parameter: atr_period = 12
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - Strategy Parameter: sl_l_period = 2.7
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - Strategy Parameter: sl_s_period = 1.3
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - Strategy Parameter: tp_l_period = 4.6
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - Strategy Parameter: tp_s_period = 5.6
2023-10-14 23:32:04,539 - freqtrade.strategy.hyper - INFO - No params for protection found, using default values.
2023-10-14 23:32:04,541 - freqtrade.data.history.history_utils - INFO - Using indicator startup period: 50 ...
2023-10-14 23:32:04,577 - freqtrade.plot.plotting - INFO - Loading data from 2023-09-01 19:50:00+00:00 to 2023-10-12 11:15:00+00:00
2023-10-14 23:32:04,578 - freqtrade.data.btanalysis - INFO - Loading backtest result from /Users/root/Desktop/source_code/freqtrade/user_data/backtest_results/backtest-result-2023-10-14_13-54-28.json
2023-10-14 23:32:04,582 - freqtrade.plot.plotting - INFO - analyse pair BTC/USDT:USDT
2023-10-14 23:32:04,829 - freqtrade.plot.plotting - WARNING - No exit_long-signals found.
2023-10-14 23:32:04,832 - freqtrade.plot.plotting - INFO - Indicator "sar" ignored. Reason: This indicator is not found in your strategy.
2023-10-14 23:32:04,832 - freqtrade.plot.plotting - INFO - Indicator "senkou_b" ignored. Reason: This indicator is not found in your strategy.
2023-10-14 23:32:04,943 - freqtrade.plot.plotting - INFO - Indicator "rsi" ignored. Reason: This indicator is not found in your strategy.
2023-10-14 23:32:04,943 - freqtrade.plot.plotting - INFO - Generate plot file for BTC/USDT:USDT
2023-10-14 23:32:05,542 - freqtrade.plot.plotting - INFO - Stored plot as /Users/root/Desktop/source_code/freqtrade/user_data/plot/freqtrade-plot-BTC_USDT_USDT-5m.html
2023-10-14 23:32:05,542 - freqtrade.plot.plotting - INFO - End of plotting process. 1 plots generated
```
# Related code
```python
    stoploss = -0.9
    use_custom_stoploss = True


def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        '''
        Omit
        '''
        true_indices = dataframe[(dataframe['highStatus'] | dataframe['lowStatus'])].index
        dataframe['sl0382'] = np.nan
        dataframe['sl0618'] = np.nan
        dataframe['sl1000'] = np.nan
        dataframe['sl1382'] = np.nan
        dataframe['sl1618'] = np.nan
        dataframe['sl2000'] = np.nan


        for i in range(2, len(true_indices)):
            a = true_indices[i - 2]
            b = true_indices[i - 1]
            c = true_indices[i]

            if dataframe['lowStatus'].iloc[a] and dataframe['highStatus'].iloc[b] and dataframe['lowStatus'].iloc[c]:
                # sl_value = dataframe['close'].iloc[a] - ((dataframe['close'].iloc[c] - dataframe['close'].iloc[b]) * 1.628)
                # sl_value = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 0.618)
                dataframe.loc[a:c, 'sl0382'] = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 0.382)
                dataframe.loc[a:c, 'sl0618'] = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 0.618)
                dataframe.loc[a:c, 'sl1000'] = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 1.000)
                dataframe.loc[a:c, 'sl1382'] = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 1.382)
                dataframe.loc[a:c, 'sl1618'] = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 1.618)
                dataframe.loc[a:c, 'sl2000'] = dataframe['close'].iloc[c] + ((dataframe['high'].iloc[b] - dataframe['close'].iloc[a]) * 2.000)
                
        dataframe['sl0382'].fillna(method='ffill', inplace=True)

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[:, 'exit_long'] = 0
        return dataframe
        
    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, after_fill: bool,
                        **kwargs) -> Optional[float]:

        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)



        last_candle = dataframe.iloc[-1].squeeze()
        
        sl0382 = last_candle['sl0382']
        
        sl0618 = last_candle['sl0618']
        
        sl1000 = last_candle['sl1000']
        
        sl1382 = last_candle['sl1382']
        
        sl1618 = last_candle['sl1618']
        
        sl2000 = last_candle['sl2000']

       
       
        if current_rate > sl0382 and current_rate < sl0618:
            return stoploss_from_open(0.07, current_profit, is_short=trade.is_short, leverage=trade.leverage)
        '''
        Omit
        '''

        return None
```
<img width="1440" alt="image" src="https://github.com/freqtrade/freqtrade/assets/22538559/99f578cd-fb04-45a2-9b7b-50a5a6ddd7ef">
 - as the picture shows. I want the current price high to be > sl0382 and less than sl0618 to initiate a trailing stop. Implement trailing stop loss one by one when the current highest price point > sl0618 < sl1000. How to implement it. Which I noticed. There is “No exit_long-signals found.” in the log. Can’t stop loss be used as an exit signal?
 - Also I want to know how to debug the custom_stoploss function in backtest mode

**Comment by stash86** [comment]
Do multiple if else. Exit long signal is only one part of exit signal. Stoploss also exit signal. The log mentioned no exit long signal.

**Comment by zhihuba** [comment]
Sorry for the trouble the description caused you. The problem is now solved.
1. Because I use "freqtrade plot-dataframe -p BTC/USDT:USDT --strategy DivergenceV5 -i 5m --timerange=20230902- ", I cannot debug.
2. Use "freqtrade backtesting" to find the appropriate parameters through debugging
## I hope you can give me some stop-profit and stop-loss suggestions from the picture.

**Comment by xmatthias** [comment]
I'm not entirely sure what you're asking / expecting.

plot-dataframe will only show what was already backtested - callbacks are NOT called or evaluated during plotting - so prior results must exist for chnages to these to show up.

**Comment by zhihuba** [comment]
All issues have been resolved. Initially I was confused because "freqtrade plot-dataframe" could not debug "custom_stoploss". When using "freqtrade backtesting" he can debug correctly
